File: Core/Features/Products/Queries/GetAllProducts.cs
namespace Core.Features.Products.Queries;

public static class GetAllProducts
{
    public record Query(ProductFilter ProductFilter) : IRequest<IEnumerable<ProductDto>>;

    public class Handler(
        IProductRepository productRepository) : IRequestHandler<Query, IEnumerable<ProductDto>>
    {
        public async Task<IEnumerable<ProductDto>> Handle(Query request, CancellationToken cancellationToken)
        {
            var products = await productRepository.GetAllAsync(
                request.ProductFilter,
                cancellationToken);

            return products.Select(p => new ProductDto(
                p.TenantCode,
                p.Name,
                p.Description,
                p.Price,
                p.TagCodes));
        }
    }

    public record ProductDto(
        int TenantCode,
        string Name,
        string? Description,
        decimal Price,
        IEnumerable<string> TagCodes,
        IEnumerable<string>? TagCategories = default!);
}

File: backoffice/src/server/queries/products/queries/get-all-products.ts
import { fetchGraphQL } from '@/utils'

export async function getAllProductsQuery(tenantCode: number) {
  const query = `
      query GetAllProducts($tenantCode: Int!) {
        allProducts(query: { productFilter: { tenantCode: $tenantCode } }) {
          name
          description
          price
          tagCodes
        }
      }
    `

  const response = await fetchGraphQL<{
    allProducts: {
      name: string
      description: string
      price: number
      tagCodes: string[]
    }[]
  }>(query, { tenantCode })

  return response.allProducts
}

export async function createProductMutation(
  tenantCode: number,
  product: {
    name: string
    description: string
    price: number
    tagCodes: string[]
  }
) {
  const mutation = `
      mutation CreateProduct(
        $tenantCode: Int!
        $name: String!
        $description: String
        $price: Decimal!
        $tagCodes: [String!]!
      ) {
        createProduct(
          command: {
            tenantCode: $tenantCode
            name: $name
            description: $description
            price: $price
            tagCodes: $tagCodes
          }
        ) {
          code
        }
      }
    `

  return await fetchGraphQL(mutation, {
    tenantCode,
    ...product,
  })
}
